### ISSUE TITLE
Fix S2583: Variable declared inline using pattern matching gives false possitive

### ISSUE THREAD

**Issue by Wouter-Codeless** [SonarSource/sonar-csharp#2028]
object a = GetValue();

if (a is string s && string.IsNullOrEmpty(s))
{
    // Do something
}

This leads to the message: Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed.

The error is related to this part of the condition "string.IsNullOrEmpty(s)"

It probably things that the variable s never gets a value, because the pattern matching syntax is not supported: "a is string s" 

Yes I know there is a work-around, this could be written different to not get this error, but I don't want to not be able to use new language features just because the analyzer doesn't understand them.

So can support for this pattern please be added? 

Visual Studio 2017.15.8.7
Sonarcloud.io
Azure Dev Ops (VSTS)

**Comment by kwilliams1987** [comment]
This appears to also manifest when using async methods which can resolve to null:

```csharp
var value = await db.DbSet.FindAsync(key);
if (value == null) {
   return HttpNotFound();
}
```

As per the documentation for `Task<TEntity> DbSet<T>.FindAsync(params Object[] keyValues)`

> If no entity is found in the context or the store, then null is returned.

**Comment by Wouter-Codeless** [comment]
Strange, according to this explanation I should get a warning to remove duplicate null checks.
But I’m getting the message: Change this condition so that it does not always evaluate to 'false';

It somehow assumes that string.IsNullOrEmpty(s) will always returning null, why is this the case?

**Comment by valhristov** [comment]
Eh, sorry, I made a mistake - `string.IsNullOrEmpty(s)` is equivalent to `s == null` for the analyzer and it indeed will say the condition is always false... I will update my previous answer.

**Comment by valhristov** [comment]
Hi @kwilliams1987, could you explain in more detail what's exactly the problem on your side? I cannot reproduce it with the snippet you sent and the latest SonarC# (7.8). A little reproducer application would be perfect for further investigation...

**Comment by valhristov** [comment]
@kwilliams1987 the rule raises in this case because the type of `userId` is Guid, which is `struct` and cannot be null, hence `userId == null` will always be false.

**Comment by kwilliams1987** [comment]
> @kwilliams1987 the rule raises in this case because the type of `userId` is Guid, which is `struct` and cannot be null, hence `userId == null` will always be false.

You make a good point, not sure how I missed that.
